---
title: "README"
author: "Timothy Nessel"
date: "10/8/2018"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

#BCB546X_R_ASSIGNMENT
##Author: Timothy Nessel

####Setting the Environment
I will start by loading the required packages if required
```{r}
if (!require("tidyverse")) install.packages("tidyverse")
if (!require("ggplot2")) install.packages("ggplot2")
if (!require("reshape2")) install.packages("reshape2")
if (!require("tidyr")) install.packages("tidyr")
if (!require("plyr")) install.packages("plyr")
library(tidyverse)
library(ggplot2)
library(reshape2)
library(tidyr)
library(plyr)
```
We will also need to load the required files from the class github repository. I specify that there is a header, and that the delimiting character is a tab. 
```{r}
fang_et_al_genotypes <- read.table("https://raw.githubusercontent.com/EEOB-BioData/BCB546X-Fall2018/master/assignments/UNIX_Assignment/fang_et_al_genotypes.txt", header = TRUE, sep = "\t")
snp_position <- read.table("https://raw.githubusercontent.com/EEOB-BioData/BCB546X-Fall2018/master/assignments/UNIX_Assignment/snp_position.txt", header = TRUE, sep = "\t")
```
####Inspecting the Files

Using the following commands, we can get some information about our two files, their dimensions, structure, column names, and the first 5 rows of the file:

```{r}
file.info("fang_et_al_genotypes")
dim(fang_et_al_genotypes)
str(fang_et_al_genotypes)
colnames(fang_et_al_genotypes)
head(fang_et_al_genotypes, n = 5)

file.info("snp_position")
dim(snp_position)
str(snp_position)
colnames(snp_position)
head(snp_position, n = 5)
```

####File Processing

We will need to create intermediate tables that only include the genotypes we are interested in, and only include the snp information we are interested in. 

To create our genotype intermediate files, we will a table with only the maize or teosinte group IDs. We will do this with which(), which will return positions when our comparison is TRUE and input them into rows.
```{r}
maize_fang <- fang_et_al_genotypes[which(fang_et_al_genotypes$Group == "ZMMIL" | fang_et_al_genotypes$Group == "ZMMLR" | fang_et_al_genotypes$Group == "ZMMMR"),]


teosinte_fang <- fang_et_al_genotypes[which(fang_et_al_genotypes$Group == "ZMPBA" | fang_et_al_genotypes$Group == "ZMPIL" | fang_et_al_genotypes$Group == "ZMPJA"),]

```

To create our snp intermediate file, we will subset our table with only columns: "SNP_ID" (,1),"Chromosome" (,3), "Position" (,4).
```{r}
snp_position_subset <- snp_position[,c(1,3,4)]
```

Our intermediate genotype files will have to be transposed in order to merge them:
Use column bind to add a new column, SNP_ID, with the row names
```{r}
maize_fang_transposed <- as.data.frame(t(maize_fang[,-1]))
colnames(maize_fang_transposed) <- maize_fang$Sample_ID
teosinte_fang_transposed <- as.data.frame(t(teosinte_fang[,-1]))
colnames(teosinte_fang_transposed) <- teosinte_fang$Sample_ID

teosinte_fang_transposed <- cbind(SNP_ID = row.names(teosinte_fang_transposed), teosinte_fang_transposed)

maize_fang_transposed <- cbind(SNP_ID = row.names(maize_fang_transposed), maize_fang_transposed)
```

We then merge the intermediate files by "SNP_ID" and row names
```{r}
merge_maize <- merge(maize_fang_transposed, snp_position, by = "SNP_ID")
merge_teosinte <- merge(teosinte_fang_transposed, snp_position, by = "SNP_ID")
```

To solve a factor error generated by "NA" data, we attempted to correct NA positions with a few strategies. Unfortunately we were unable to solve this error, leading to a warning suppression later on.
```{r}
#merge_maize$Position <- as.numeric(as.character(merge_maize$Position))
#merge_maize$Chromosome <- as.numeric(as.character(merge_maize$Chromosome))
#merge_maize <- merge_maize[!(merge_maize$Position == "unknown")]
#merge_maize <- merge_maize[!(merge_maize$Chromosome == "unknown")]
#merge_maize <- merge_maize[!(merge_maize$Chromosome == "multiple")]
#merge_maize <- merge_maize[!(is.na(merge_maize$Position))]
#merge_maize[merge_maize=="N/A"] <- "?/?"
#merge_maize[is.na(merge_maize)] <- "?/?"
#merge_teosinte[merge_teosinte == "N/A"] <- "?/?"
#merge_teosinte[is.na(merge_teosinte)] <- "?/?"
#merge_maize_reverse[is.na(merge_maize_reverse)] <- "?/?"
#merge_teosinte_reverse[is.na(merge_teosinte_reverse)] <- "?/?"
```

To sort our merged files by increasing and decreasing position, we use the order() function. We also replace 'NA' data with ?/?:
```{r}
merge_maize <- merge_maize[order(merge_maize$Position),]
merge_teosinte <- merge_teosinte[order(merge_teosinte$Position),]
merge_maize_reverse <- merge_maize[order(merge_maize$Position,decreasing = T),]
merge_teosinte_reverse <- merge_teosinte[order(merge_teosinte$Position,decreasing = T),]
```

Then we use for loops to subset our datasets by chromosome and write to new files
```{r}
for (i in 1:10){
  merge_maize_chromosome <- merge_maize[merge_maize$Chromosome == i,]
  write.csv(merge_maize_chromosome, file=paste("maize_increasing_",i,".csv",sep=""))
}
for (i in 1:10){
  merge_maize_chromosome_rev <- merge_maize_reverse[merge_maize_reverse$Chromosome == i,]
  suppressWarnings(merge_maize_chromosome_rev[merge_maize_chromosome_rev == "?/?"] <- factor("-/-"))
  write.csv(merge_maize_chromosome_rev, file=paste("maize_decreasing_",i,".csv",sep=""))
}
for (i in 1:10){
  merge_teosinte_chromosome <- merge_teosinte[merge_teosinte$Chromosome == i,]
  write.csv(merge_teosinte_chromosome, file=paste("teosinte_increasing_",i,".csv",sep=""))
}
for (i in 1:10){
  merge_teosinte_chromosome_rev <- merge_teosinte_reverse[merge_teosinte_reverse$Chromosome == i,]
  suppressWarnings(merge_teosinte_chromosome_rev[merge_teosinte_chromosome_rev == "?/?"] <- factor("-/-"))
  write.csv(merge_teosinte_chromosome_rev, file=paste("teosinte_decreasing_",i,".csv",sep=""))
}
```

####Visualizing the Data

To plot the total number of SNPs in our dataset on each chromosome, we need to create a merged dataframe with both all groups, so we need to repeat some earlier steps.
```{r}
fang_transposed <- as.data.frame(t(fang_et_al_genotypes[,-1]))
fang_transposed <- cbind(SNP_ID = row.names(fang_transposed), fang_transposed)
merge_fang <- merge(fang_transposed, snp_position, by = "SNP_ID")
merge_fang$Chromosome <- factor(merge_fang$Chromosome)
ggplot(merge_fang, aes((Chromosome))) + geom_bar()
```

To see what groups contribute the most SNPs, we can do a simple plot with the original data.
```{r}
ggplot(fang_et_al_genotypes, aes(Group)) + geom_bar()
```
Groups that contribute the most are: #1 ZMMLR, #2 ZMMIL, #3 ZMPBA

To plot missing data and heterozygosity, we need to first process the files to show these data. We can create a new column that records whether the genotype characters match. In all of fang_et_al_genotypes, we can see a general proportion of homozygous to heterozygous to missing data. 
```{r}
fang_melted <- melt(fang_et_al_genotypes, id = c("Sample_ID", "Group"))
fang_melted[fang_melted == "?/?" ] = NA
fang_melted$Homozygous <- (fang_melted$value=="A/A" | fang_melted$value=="C/C" | fang_melted$value=="G/G" | fang_melted$value=="T/T")
ggplot(fang_melted, aes(Homozygous)) + geom_bar()
```

To group this data by the species and group we will make some new intermediate data frames.
```{r}
fang_melted_group <- fang_melted[order(fang_melted$Group),]
fang_melted_ID <- fang_melted[order(fang_melted$Sample_ID),]
```

To get the proportion of homo/heterozygosity for each ordered dataframe, we have to summarize the 'Homozygous' column we made. I had to get help from more senior BCB students for this section. 
```{r}
fang_melted_group_prop <- ddply(fang_melted_group, c("Group"), summarise, homo=sum(Homozygous, na.rm=TRUE), 
                     heter=sum(Homozygous, na.rm=FALSE), NotValid=sum(is.na(Homozygous)))
group_prop_melt <- melt(fang_melted_group_prop, measure.vars = c("homo", "heter", "NotValid"))

fang_melted_ID_prop <- ddply(fang_melted_ID, c("Sample_ID"), summarise, homo=sum(Homozygous, na.rm=TRUE), 
                     heter=sum(Homozygous, na.rm=FALSE), NotValid=sum(is.na(Homozygous)))
ID_prop_melt <- melt(fang_melted_ID_prop, measure.vars = c("homo", "heter", "NotValid"))

ggplot(group_prop_melt,aes(x = Group, y= value, fill=variable)) + geom_bar(stat = "identity", position = "stack")
ggplot(ID_prop_melt,aes(x = Sample_ID, y= value, fill=variable)) + geom_bar(stat = "identity", position = "stack")
```

For my own graph, I wanted to calculate percentage of A and T homozygotes as opposed to G and C homozygotes ordered by groups, in an albeit similiar style as the last example. 
```{r}
fang_melted$AT <- (fang_melted$value=="A/A" | fang_melted$value=="T/T")
fang_melted$GC <- (fang_melted$value=="G/G" | fang_melted$value=="C/C")
fang_melted_group <- fang_melted[order(fang_melted$Group),]
fang_melted_AT_GC <- ddply(fang_melted_group, c("Group"), summarise, A_or_T_homozygous=sum(AT, na.rm=TRUE), 
                     G_or_C_homozygous=sum(GC, na.rm=TRUE))
fang_melted_AT_GC_melt <- melt(fang_melted_AT_GC, measure.vars = c("A_or_T_homozygous", "G_or_C_homozygous"))
ggplot(fang_melted_AT_GC_melt,aes(x = Group, y= value, fill=variable)) + geom_bar(stat = "identity", position = "stack")
```



